### PR TITLE
bpo-42064: Move sqlite3 types to global state

### PR DESCRIPTION
**Pull request by erlend-aasland**
https://bugs.python.org/issue42064

### CODE REVIEW

**Comment by erlend-aasland** [comment]
Internal change only; skipping news.

File: Modules/_sqlite/microprotocols.c [hunk 1]
```diff
@@ -58,8 +58,8 @@ pysqlite_microprotocols_add(PyTypeObject *type, PyObject *proto, PyObject *cast)
     PyObject* key;
     int rc;
 
-    if (proto == NULL) proto = (PyObject*)pysqlite_PrepareProtocolType;
-
+    assert(type != NULL);
+    assert(proto != NULL);
```
Comment (erlend-aasland) [inline Modules/_sqlite/microprotocols.c:62]: There is only one user of `pysqlite_microprotocols_add`; `proto` will never be `NULL`.
https://github.com/python/cpython/blob/2780df478129c4f1042df2868b6adf2bf507cd6e/Modules/_sqlite/module.c#L179

Converting the sanity check to an assert instead.

**Comment by erlend-aasland** [comment]
@vstinner or @encukou: could one of you take a look at this? It's mostly search-and-replace (ish), except for the one case I've pointed out. Let me know if I should break this up in multiple PR's.

**Comment by encukou** [comment]
Hmm, I'm not too familiar with the plan for the sqlite module. The calls to `pysqlite_get_state(NULL)` will be replaced with non-NULL in the future, right?

It seems that so far, the only module that uses dynamic types with Argument Clinic's `subclass_of` is `_ssl`, and AFAICS, it does things by walking the MRO twice. It might be worth it to add better support for this need. For example, I could imagine checking if any superclass was defined using a given *PyType_Spec*.

**Comment by encukou** [comment]
OK, then this PR looks like a good step!

> You're talking about a new API to optimise this, right?
Yes. I noted it and will try to make some time for it.

**Comment by erlend-aasland** [comment]
> OK, then this PR looks like a good step!

Yes, I believe it is. I think it's best to keep the "establish global state" PR's as simple as possible, for the reviewers convenience. Optimisations, for example passing state as a parameter to a function, or storing state members in the object context (as `functools` does), are easier add afterwards, IMO.

> > You're talking about a new API to optimise this, right?
> Yes. I noted it and will try to make some time for it.

Fantastic. I'd be interested in following that process, for the sake of learning.

**Comment by erlend-aasland** [comment]
LGTY, @encukou?

**Comment by encukou** [comment]
Looks good; thanks for the clarification!

**Comment by erlend-aasland** [comment]
Thanks for reviewing, @encukou !

**Comment by encukou** [comment]
> Fantastic. I'd be interested in following that process, for the sake of learning.

I don't have time to get into this, but I put down some notes: https://github.com/encukou/abi3/issues/19

**Pull request merged**